src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1013 @@
+                chunk.meshId = mesh->typed_id;
+                chunk.meshMaterialIndex = (UnsignedInt)i;
+                ++chunkOffset;
+            }
+        }
+    }
+
+    /* Count the final number of nodes in the scene, we may remove some (root)
+       or add (synthetic geometry transform nodes) */
+    {
+        const bool preserveRootNode = configuration().value<bool>("preserveRootNode");
+        const bool geometricTransformNodes = configuration().value<bool>("geometricTransformNodes");
+
+        _state->nodeIdOffset = 0;
+        _state->originalNodeCount = (UnsignedInt)scene->nodes.count;
+
+        if(!preserveRootNode) {
+            --_state->originalNodeCount;
+            ++_state->nodeIdOffset;
+        }
+
+        _state->nodeCountWithSynthetic = _state->originalNodeCount;
+
+        /* Reserve space for nodes if we want to create dummy nodes for geometric
+           transforms */
+        if(geometricTransformNodes) {
+            for(ufbx_node *node : scene->nodes) {
+                if(node->has_geometry_transform) {
+                    ++_state->nodeCountWithSynthetic;
+                }
+            }
+        }
+    }
+}
+
+Int UfbxImporter::doDefaultScene() const { return 0; }
+
+UnsignedInt UfbxImporter::doSceneCount() const { return 1; }
+
+Int UfbxImporter::doSceneForName(Containers::StringView) {
+    return -1;
+}
+
+Containers::String UfbxImporter::doSceneName(UnsignedInt) {
+    return {};
+}
+
+Containers::Optional<SceneData> UfbxImporter::doScene(UnsignedInt) {
+    ufbx_scene *scene = _state->scene.get();
+
+    const bool preserveRootNode = configuration().value<bool>("preserveRootNode");
+    const bool geometricTransformNodes = configuration().value<bool>("geometricTransformNodes");
+
+    UnsignedInt meshCount = 0;
+    UnsignedInt skinCount = 0;
+    UnsignedInt nodeCount = _state->nodeCountWithSynthetic;
+    UnsignedInt cameraCount = 0;
+    UnsignedInt lightCount = 0;
+
+    /* ufbx meshes can contain per-face materials so we need to separate them
+       into pieces containing a single material for SceneData. */
+    for(std::size_t i = 0; i < scene->meshes.count; ++i) {
+        ufbx_mesh *mesh = scene->meshes[i];
+
+        UnsignedInt instanceCount = (UnsignedInt)mesh->instances.count;
+        for(const ufbx_mesh_material &mat : mesh->materials) {
+            if(mat.num_faces == 0) continue;
+            meshCount += instanceCount;
+            if(mesh->skin_deformers.count > 0)
+                skinCount += instanceCount;
+        }
+    }
+
+    /* Collect instanced camera/light counts */
+    for(ufbx_light *light : scene->lights)
+        lightCount += (UnsignedInt)light->instances.count;
+    for(ufbx_camera *camera : scene->cameras)
+        cameraCount += (UnsignedInt)camera->instances.count;
+
+    /* Allocate the output array. */
+    Containers::ArrayView<UnsignedInt> nodeObjects;
+    Containers::ArrayView<const void*> importerState;
```

Comment:
The importer state pointers are only useful when the types they point to are accessible (such as, in case of Assimp being an external library, pointing to an `aiNode` and the user can include the corresponding header and cast to that type). Here ufbx is bundled inside the dynamically loaded plugin which prevents that, plus IIUC you `ufbx_free(scene)` at the end of `doOpen*()` so the pointer is dangling anyway?

I'd just drop the `importerState` field altogether for now, and similarly for the `importerState` constructor parameter in all `Trade::*Data` classes.

Reply:
I thought that users could "bring their own" ufbx header and interpret the types but maybe it's a pain to keep the versions in sync, so could just get rid of all that. The lifetimes are fine as the scene is stored in `ufbx_scene_ref` that keeps it alive in C++ RAII style, maybe it's confusing to have that kind of C++ism in a C library.

Reply:
If it becomes useful / needed at some point, I can add that, but until then it's just extra maintenance overhead. Remove, pls :)

Reply:
True, if users need full control of ufbx they can just `#include "ufbx.h"` and go from there :)